File: src/bumperbot_utils/bumperbot_utils/battery_level_pub.py
#!/usr/bin/env python3
import time
import smbus

import rclpy
from rclpy.node import Node

from std_msgs.msg import Float32


# Config Register (R/W)
_REG_CONFIG = 0x00

# SHUNT VOLTAGE REGISTER (R)
_REG_SHUNTVOLTAGE = 0x01

# BUS VOLTAGE REGISTER (R)
_REG_BUSVOLTAGE = 0x02

# POWER REGISTER (R)
_REG_POWER = 0x03

# CURRENT REGISTER (R)
_REG_CURRENT = 0x04

# CALIBRATION REGISTER (R/W)
_REG_CALIBRATION = 0x05


class BusVoltageRange:
    RANGE_16V = 0x00
    RANGE_32V = 0x01


class Gain:
    DIV_1_40MV = 0x00
    DIV_2_80MV = 0x01
    DIV_4_160MV = 0x02
    DIV_8_320MV = 0x03


class ADCResolution:
    ADCRES_9BIT_1S = 0x00
    ADCRES_12BIT_32S = 0x0D


class Mode:
    SANDBVOLT_CONTINUOUS = 0x07


class INA219:
    def __init__(self, i2c_bus=7, addr=0x41):
        self.bus = smbus.SMBus(i2c_bus)
        self.addr = addr

        self._cal_value = 0
        self._current_lsb = 0
        self._power_lsb = 0

        self.set_calibration_16V_5A()

    def read(self, address):
        data = self.bus.read_i2c_block_data(self.addr, address, 2)
        return (data[0] << 8) + data[1]

    def write(self, address, data):
        self.bus.write_i2c_block_data(
            self.addr,
            address,
            [(data >> 8) & 0xFF, data & 0xFF]
        )

    def set_calibration_16V_5A(self):

        self._current_lsb = 0.1524
        self._cal_value = 26868
        self._power_lsb = 0.003048

        self.write(_REG_CALIBRATION, self._cal_value)

        config = (
            (BusVoltageRange.RANGE_16V << 13)
            | (Gain.DIV_2_80MV << 11)
            | (ADCResolution.ADCRES_12BIT_32S << 7)
            | (ADCResolution.ADCRES_12BIT_32S << 3)
            | Mode.SANDBVOLT_CONTINUOUS
        )

        self.write(_REG_CONFIG, config)

    def getShuntVoltage_mV(self):
        value = self.read(_REG_SHUNTVOLTAGE)

        if value > 32767:
            value -= 65536

        return value * 0.01

    def getBusVoltage_V(self):
        value = self.read(_REG_BUSVOLTAGE)
        return (value >> 3) * 0.004

    def getCurrent_mA(self):
        value = self.read(_REG_CURRENT)

        if value > 32767:
            value -= 65536

        return value * self._current_lsb

    def getPower_W(self):
        value = self.read(_REG_POWER)

        if value > 32767:
            value -= 65536

        return value * self._power_lsb


class INA219Publisher(Node):

    def __init__(self):
        super().__init__("ina219_publisher")

        # Parameters
        self.declare_parameter("i2c_bus", 7)
        self.declare_parameter("i2c_addr", 0x41)
        self.declare_parameter("publish_rate", 2.0)

        i2c_bus = self.get_parameter("i2c_bus").value
        i2c_addr = self.get_parameter("i2c_addr").value
        publish_rate = self.get_parameter("publish_rate").value

        # INA219
        self.ina219 = INA219(
            i2c_bus=i2c_bus,
            addr=i2c_addr
        )

        # Publishers
        self.current_pub = self.create_publisher(
            Float32,
            "/current",
            10
        )

        self.percentage_pub = self.create_publisher(
            Float32,
            "/percentage",
            10
        )

        self.load_voltage_pub = self.create_publisher(
            Float32,
            "/load_voltage",
            10
        )

        self.power_pub = self.create_publisher(
            Float32,
            "/power",
            10
        )

        # Timer
        self.timer = self.create_timer(
            1.0 / publish_rate,
            self.publish_data
        )

        self.get_logger().info("INA219 ROS2 Publisher Started")

    def publish_data(self):

        try:
            bus_voltage = self.ina219.getBusVoltage_V()
            shunt_voltage = self.ina219.getShuntVoltage_mV() / 1000.0
            current = self.ina219.getCurrent_mA() / 1000.0
            power = self.ina219.getPower_W()

            percentage = (bus_voltage - 9) / 3.6 * 100
            percentage = max(0.0, min(percentage, 100.0))

            # Create messages
            current_msg = Float32()
            current_msg.data = float(current)

            percentage_msg = Float32()
            percentage_msg.data = float(percentage)

            load_voltage_msg = Float32()
            load_voltage_msg.data = float(bus_voltage)

            power_msg = Float32()
            power_msg.data = float(power)


            # Publish
            self.current_pub.publish(current_msg)
            self.percentage_pub.publish(percentage_msg)
            self.load_voltage_pub.publish(load_voltage_msg)
            self.power_pub.publish(power_msg)
            self.get_logger().info(
                f"Voltage: {bus_voltage:.2f} V | "
                f"Current: {current:.2f} A | "
                f"Power: {power:.2f} W | "
                f"Battery: {percentage:.1f}%"
            )

        except Exception as e:
            self.get_logger().error(f"INA219 Read Error: {e}")


def main(args=None):

    rclpy.init(args=args)

    node = INA219Publisher()

    try:
        rclpy.spin(node)

    except KeyboardInterrupt:
        pass

    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/bumperbot_utils/src/battery_level_pub.cpp
#include <iostream>
#include <iomanip>
#include <cstdint>
#include <unistd.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <linux/i2c-dev.h>

#define REG_CONFIG         0x00
#define REG_SHUNTVOLTAGE   0x01
#define REG_BUSVOLTAGE     0x02
#define REG_POWER          0x03
#define REG_CURRENT        0x04
#define REG_CALIBRATION    0x05

class INA219 {
private:
    int file;
    uint8_t addr;

    uint16_t cal_value;
    float current_lsb;
    float power_lsb;

public:
    INA219(const char* i2c_device = "/dev/i2c-7", uint8_t address = 0x40) {
        addr = address;

        file = open(i2c_device, O_RDWR);
        if (file < 0) {
            perror("Failed to open I2C bus");
            exit(1);
        }

        if (ioctl(file, I2C_SLAVE, addr) < 0) {
            perror("Failed to connect to INA219");
            exit(1);
        }

        setCalibration16V5A();
    }

    ~INA219() {
        close(file);
    }

    uint16_t readRegister(uint8_t reg) {
        uint8_t buffer[2];

        if (write(file, &reg, 1) != 1) {
            perror("Failed to write register address");
        }

        if (read(file, buffer, 2) != 2) {
            perror("Failed to read register");
        }

        return (buffer[0] << 8) | buffer[1];
    }

    void writeRegister(uint8_t reg, uint16_t value) {
        uint8_t buffer[3];

        buffer[0] = reg;
        buffer[1] = (value >> 8) & 0xFF;
        buffer[2] = value & 0xFF;

        if (write(file, buffer, 3) != 3) {
            perror("Failed to write register");
        }
    }

    void setCalibration16V5A() {
        current_lsb = 0.1524f;
        cal_value = 26868;
        power_lsb = 0.003048f;

        writeRegister(REG_CALIBRATION, cal_value);

        uint16_t config =
            (0x00 << 13) |   // 16V range
            (0x01 << 11) |   // Gain /2, 80mV
            (0x0D << 7)  |   // 12-bit ADC, 32 samples
            (0x0D << 3)  |   // 12-bit ADC, 32 samples
            (0x07);          // Continuous mode

        writeRegister(REG_CONFIG, config);
    }

    float getShuntVoltage_mV() {
        int16_t value = (int16_t)readRegister(REG_SHUNTVOLTAGE);
        return value * 0.01f;
    }

    float getBusVoltage_V() {
        uint16_t value = readRegister(REG_BUSVOLTAGE);
        return ((value >> 3) * 0.004f);
    }

    float getCurrent_mA() {
        int16_t value = (int16_t)readRegister(REG_CURRENT);
        return value * current_lsb;
    }

    float getPower_W() {
        int16_t value = (int16_t)readRegister(REG_POWER);
        return value * power_lsb;
    }
};

int main() {
    INA219 ina219("/dev/i2c-7", 0x41);

    while (true) {
        float bus_voltage = ina219.getBusVoltage_V();
        float shunt_voltage = ina219.getShuntVoltage_mV() / 1000.0f;
        float current = ina219.getCurrent_mA();
        float power = ina219.getPower_W();

        float percentage = (bus_voltage - 9.0f) / 3.6f * 100.0f;

        if (percentage < 0)
            percentage = 0;

        if (percentage > 100)
            percentage = 100;

        std::cout << std::fixed << std::setprecision(3);

        std::cout << "Load Voltage:  "
                  << bus_voltage
                  << " V" << std::endl;

        std::cout << "Current:       "
                  << current / 1000.0f
                  << " A" << std::endl;

        std::cout << "Power:         "
                  << power
                  << " W" << std::endl;

        std::cout << "Percentage:    "
                  << std::setprecision(2)
                  << percentage
                  << " %" << std::endl;

        std::cout << "-----------------------------"
                  << std::endl;

        sleep(2);
    }

    return 0;
}